firefox-pmos-mobile: init at -pmos-2.2.0

zhaofengli opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

This PR adds a wrapped version of Firefox with postmarketOS's [mobile-friendly configurations](https://gitlab.com/postmarketOS/mobile-config-firefox). The package version is `(browser version)-pmos-2.0.1`. It's possible to remove the postmarketOS-branded initial homepage by setting `keepHomepage = true;`.


<details>
<summary>Screenshot</summary>
<img src="https://user-images.githubusercontent.com/2189609/116769772-9cef2f00-a9f3-11eb-85cf-e19502db634a.jpg" alt="Screenshot" style="max-width:30%;">
</details>

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

zhaofengli committed:
firefox-pmos-mobile: init at -pmos-2.2.0

samueldr commented:
May I suggest we drop the postmarketOS branded homepage by default?

I'm also not sure what are the guidelines with Mozilla branding and non-default homepages.

zhaofengli commented:
> May I suggest we drop the postmarketOS branded homepage by default?

Sure, I guess it makes things less confusing.

> I'm also not sure what are the guidelines with Mozilla branding and non-default homepages.

That's actually a good point, I'm not sure if we can use the Firefox trademark with such modifications sans approval from Mozilla.

samueldr commented:
Reminder to committers/contributors: don't remove the ofborg internal error tag! Only a collective member can do so safely!

* * *

I don't know if changes that are not changing the "trademarky" bits are or are not fine. And even me saying it might not be fine to change the homepage is entirely conjecture.

dotlambda commented on pkgs/applications/networking/browsers/firefox-pmos-mobile/default.nix:
```diff
@@ -0,0 +1,61 @@
+    '';
+
+    fixupPhase = ''
+      sed -i "s|/etc/mobile-config-firefox/userChrome.css|$out/userChrome.css|" $out/mobile-config-autoconfig.js
+      sed -i "s|/usr/share/mobile-config-firefox/home.html|$out/home.html|" $out/policies.json
```
Use `substituteInPlace`.

dotlambda reviewed (commented):
We can use `makeFlags = [ "DISTRO=NixOS" ]` to at least not have the homepage show "postmarketOS".

Pacman99 commented on pkgs/applications/networking/browsers/firefox-pmos-mobile/default.nix:
```diff
@@ -0,0 +1,61 @@
+    };
+  };
+
+  # AutoConfig (forced configs)
+  extraPrefs = builtins.readFile "${mobile-config-firefox}/mobile-config-autoconfig.js";
+
+  extraPolicies = let
+    policies = (builtins.fromJSON (builtins.readFile "${mobile-config-firefox}/policies.json")).policies;
+  in if keepHomepage then lib.filterAttrs (k: v: k != "Homepage") policies else policies;
```
Isn't this IFD? I thought nixpkgs and hydra was against that.
Maybe the files could be copied into the tree

zhaofengli replied:
Sure, let's have a update script that generates a `.nix`.

zhaofengli committed:
wrapFirefox: Make the desktop name suffix configurable for forceWayland

This is useful for hiding the suffix when Wayland is the only usable
display protocol on the system (e.g., Phosh without XWayland).

dotlambda commented:
Why don't we call the package `mobile-config-firefox`? That's what the AUR and postmarketOS do: https://repology.org/project/mobile-config-firefox/versions.

dotlambda commented on pkgs/applications/networking/browsers/firefox-pmos-mobile/default.nix:
```diff
@@ -0,0 +1,70 @@
+      mkdir $out
+      cp out/home.html $out/
+      cp out/userChrome.css $out/
+      cp src/policies.json $out/
```
```suggestion
```

dotlambda commented on pkgs/applications/networking/browsers/firefox-pmos-mobile/default.nix:
```diff
@@ -0,0 +1,70 @@
+  version = "2.0.1";
+
+  mobile-config-firefox = stdenv.mkDerivation {
+    name = "mobile-config-firefox";
```
```suggestion
    pname = "mobile-config-firefox";
    version = "2.0.1";

```

dotlambda commented on pkgs/applications/networking/browsers/firefox-pmos-mobile/default.nix:
```diff
@@ -0,0 +1,70 @@
+      sha256 = "0kqq31h4qdqbn8y65jz27rsn7v85q3bdprdx1fb0c9s3g9lh5zqx";
+    };
+
+    makeFlags = lib.optional (distroName != null) "DISTRO=${distroName}";
```
I don't see a reason to make this overridable. If people want to change it, they should use `overrideAttrs`.

dotlambda commented on pkgs/applications/networking/browsers/firefox-pmos-mobile/default.nix:
```diff
@@ -0,0 +1,70 @@
+{ browser
```
```suggestion
{ firefox
```
No other browser will work.

dotlambda reviewed (changes requested):
We should add a check for version consistency between `policies.nix` and `default.nix` using `assert`.

zhaofengli commented:
@dotlambda 

> Why don't we call the package `mobile-config-firefox`? That's what the AUR and postmarketOS do: https://repology.org/project/mobile-config-firefox/versions.

It's the config itself rather than the wrapped browser with the `firefox` executable. Here I call it `mobile-config-firefox` as well but don't expose it in `all-packages` as it's not meant to be consumed alone.

zhaofengli commented:
Gentle poke. Can we get this merged?

SuperSandro2000 commented on pkgs/applications/networking/browsers/firefox-pmos-mobile/generate.sh:
```diff
@@ -0,0 +1,17 @@
+
+set -euo pipefail
+
+if [ "$#" -ne 1 ] || [[ "$1" == -* ]]; then
```
```suggestion
if [[ "$#" -ne 1 || "$1" == -* ]]; then
```

uninsane commented:
@zhaofengli i rebased your patches against master, updated 2.2.0 -> 4.0.2, and implemented an updateScript. deployed to a Pinephone, and it's working as expected.

if you care to shepherd this PR through, feel free to pull my patches. i've left them unsquashed so that you can see what changes -- probably want to squash them before merging. i implemented `librewolf-pmos-mobile` to show that we got the `libName` stuff right -- probably not worth shipping it though: https://github.com/NixOS/nixpkgs/compare/master...uninsane:nixpkgs:firefox-pmos-mobile

zhaofengli commented:
@uninsane Unfortunately I haven't touched mobile stuff in a while so it would take me some time to ramp back up and test things. Could you open a new PR and link it here? I'll then close this one.

bendlas commented:
@uninsane at https://github.com/NixOS/nixpkgs/pull/206054#issuecomment-1608640136
> @bendlas if you have a moment, care to take a look over https://github.com/NixOS/nixpkgs/pull/121356#issuecomment-1608619585 to weigh in? i just posted a patch there that gets us inline with upstream, but it's a pretty different approach than what just got merged, so i'm not sure the best way to proceed.

Hey, sorry, I didn't see we had this, before merging. I've got no problem with updating what I merged, with this PR (including the name, `firefox-pmos-mobile`), if you're inclined to rebase.

I'll inline comment some points:

bendlas commented on pkgs/applications/networking/browsers/firefox-pmos-mobile/default.nix:
```diff
@@ -0,0 +1,70 @@
+    sed '/general\.config\.\(filename\|obscure_value\)/ s|^|//|g' < "${mobile-config-firefox}/mobile-config-prefs.js" > "$out/lib/${firefoxLibName}/defaults/pref/mobile-config-prefs.js"
+
+    # Inject forced configs
+    cat "${mobile-config-firefox}/mobile-config-autoconfig.js" >> "$out/lib/${firefoxLibName}/mozilla.cfg"
```
I guess sedding out the auto-config from the upstream prefs.js is a valid alternative to what I did, which is building up our own prefs.js

Both are brittle in their own way, but this here might be slightly less maintenance burden.

The main advantage I see for my packaging is, that I'm copying out referenced files (userChrome.css and userContent.css), so the source package isn't referred to at runtime any more, but that might be outweighed by the more frequent need to compare upstream manually.

bendlas reviewed (commented):
As I said, overall I'm fine with taking this approach.

uninsane commented:
the patch i shared above causes firefox to ignore touch input. i'm not sure if this is an issue only when using wayland, or sway, or if it's a packaging bug. either way, i'm finding it more practical to use webkit-based browsers on mobile (Tangram, epiphany), so putting this on hold.

but if anyone else makes progress on updating this package, ping me & i'll be happy to test it on hardware for you.
